Autotune docs (#1661)
PX4/PX4-user_guide: en/config/autotune.md

Generally if you use an [airframe configuration](../config/airframe.md) that is similar to your vehicle then the vehicle will be able to fly.
However unless the configuration precisely matches your hardware you should tune the rate and attitude controllers.
Tuning the velocity and position controllers is less important because they are less affected by vehicle dynamics, and the default tuning configuration for a similar airframe is often sufficient.

Autotuning provides an automatic mechanism to tune the rate and attitude controllers.
It can be used to tune fixed wing and multicopter vehicles, and VTOL vehicles when flying as a multicopter or as a fixed wing (transition between modes must be manually tuned).
In theory it should work for other vehicle types that have a rate controller, but currently only the above types are supported.

Automatic tuning works well for the multicopter and fixed wing vehicle configurations supported by PX4, provided the frame is not too flexible
(see [below for more information](#does-autotuning-work-for-all-supported-airframes)).

> The vehicle must be flying in [Hold mode](../flight_modes/hold.md), [Position mode](../flight_modes/position_mc.md), or [altitude mode](../flight_modes/altitude_mc.md).

Author: Is this true? I.e. can I tune in "hold mode"? Does it matter. The original docs appear to indicate that you can't assume that position hold mode will work and are written that way. Should I do so too?

Reviewer: You can and I would recommend tuning in hold or position mode if the drone can already fly in those modes with the default tuning (which should be the case if it passes the `#pre-tuning-test`). In the original doc, we wanted to avoid having a user performing an automatic takeoff without knowing if the drone was able to fly.

The flight stack will apply a small disturbance to the vehicle in each axis and then attempt to calculate the new tuning parameters.
For fixed wing vehicles the new tuning is applied in-air by default, after which the vehicle tests the new settings and reverts the tuning if the controllers are not stable.

> For multicopter, the vehicle lands and applies the new tuning parameters after disarming; the pilot is expected to then take off carefully and test the tuning.

Reviewer: The user needs to land the drone himself (trigger landing or land manually).

The tuning process takes about 40 seconds ([between 19 and 68 seconds](#how-long-does-autotuning-take)).
The default behaviour can be configured using [parameters](#parameters).

:::note
For instructions on using the manual PID tuning see:

> - [Multicopter PID Tuning Guide](../config_mc/pid_tuning_guide_multicopter_basic.md) (Manual/Simple)

Author: I am assuming these are still useful. BUT we do need someone else - @bresch to go through and update. Unless this is just a matter of adding a heading "you can probably do this with autotune" at the top (can I do this)?

For advanced tuning, what is still useful? Just the stuff around airmode?

The point is, we want to point to what you do next, and tell people what they can now ignore.

- [Multicopter PID Tuning Guide](../config_mc/pid_tuning_guide_multicopter.md) (Advanced/Detailed)
- [Fixed-Wing PID Tuning Guide](../config_fw/pid_tuning_guide_fixedwing.md)
:::

<!-- 
Overview of the auto-tuning workflow

@[youtube](https://youtu.be/uHVlIIydcpg)
-->

## Pre-tuning Test

> The vehicle must be able to fly and adequately stabilize itself before running auto-tune.

Author: ```suggestion
The vehicle must be able to fly and adequately stabilize itself before running auto-tune.
This test ensures that the vehicle can fly safely in position controlled modes..
```

:::note
During [Airframe Setup](../config/airframe.md) you should have selected the frame that most closely matches your vehicle.
This will usually be tuned well enough to fly, and it _may_ also be sufficiently well tuned to run autotuning.
:::

To make sure the vehicle is stable enough for auto-tuning:

1. Perform a normal preflight safety checklist to ensure the flight zone is clear and has enough space.
2. Takeoff and prepare for the test

>    - **Multicopters:** Take off and hover at 1m above ground in Position mode, Altitude mode or Hold mode.

Reviewer: ```suggestion
   - **Multicopters:** Take off and hover at 1m above ground in Altitude or Stabilized mode.
```

Those tests are mainly to see if the drone can fly safely in position in auto modes by checking the attitude stability so they need to be performed in stabilized or altitude modes.

>    - **Fixed-wing:** Take off and fly at cruise speed in Position mode or Alitude mode.

Reviewer: ```suggestion
   - **Fixed-wing:** Take off and fly at cruise speed in Altitude mode or Stabilized mode.
```

   A above, start with small angles and confirm that the vehicle can itself within 2 oscillations before increasing the tilt.

If the drone can stabilize itself within 2 oscillations it is ready for the auto-tuning procedure.

If not, go to the [troubleshooting](#troubleshooting) section, which explains the minimal manual tuning to prepare the vehicle for auto-tuning.

### Auto-tuning procedure

The auto-tuning sequence must be performed in a **safe flight zone, with enough space**.
For best results, it is recommended to fly in calm weather conditions.

:::note
The sequence can be aborted at any time by the operator by moving the roll/pitch stick on the RC controller.
:::

1. Perform the [pre-tuning test](#pre-tuning-test) detailed previously
3. Takeoff using RC control and prepare for test:
   - **Multicopters:** Takeoff using the remote controller in altitude mode.
     Hover the vehicle at a safe distance and at a few meters above ground (between 4 and 20m).
     :::note
     If the drone is known to be stable in position controlled mode, you can 
     - run auto-tuning in position mode and also takeoff using takeoff mode.
     - use takeoff mode to take off.
     :::
      

>    - **Fixed-wing:** Once flying at cruise speed, activate auto mode.

Author: Do you mean hold mode? Auto is "mission".

Reviewer: Yes, hold mode

Author: ```suggestion
   - **Fixed-wing:** Once flying at cruise speed, activate [Hold mode](../flight_modes/hold.md).
```

This behaviour can be configured using the [MC_AT_APPLY](../advanced_config/parameter_reference.md#MC_AT_APPLY) and [FW_AT_APPLY](../advanced_config/parameter_reference.md#FW_AT_APPLY) parameters respectively:

* `0`: the gains are not applied.
  This is used for testing purposes if the user wants to inspect results of the auto-tuning algorithm without using them directly.
* `1`: apply the gains after disarm (default for multirotors).
  The operator can then test the new tuning while taking-off carefully.
* `2`: apply immediately (default for fixed-fings).
  The new tuning is applied, disturbances are sent to the controller and the stability is monitored during the next 4 seconds.
  If the control loop is unstable, the control gains are immediately reverted back to their previous value.
  If the test passes, the pilot can then use the new tuning.

Fixed wing vehicles (only) can tune which axes are tuned using the [FW_AT_AXES](../advanced_config/parameter_reference.md#FW_AT_AXES) bitmask parameter:

* bit `0`: roll (default)
* bit `1`: pitch (default)
* bit `2`: yaw

### Troubleshooting

> #### The drone oscillates when performing the testing maneuvers prior to the multirotor auto-tuning

Reviewer: ```suggestion
#### The drone oscillates when performing the testing maneuvers prior to the auto-tuning
```

> * slow oscillations (1 oscillation per second or slower): this often occurs on large platforms and means that the attitude loop is too fast compared to the rate loop -> decrease `MC_[ROLL|PITCH]_P` by steps of 1.0.

Reviewer: ```suggestion
* slow oscillations (1 oscillation per second or slower): this often occurs on large platforms and means that the attitude loop is too fast compared to the rate loop.
   - **Multicopter:** decrease `MC_[ROLL|PITCH]_P` by steps of 1.0
   - **Fixed-wing:** increase `FW_RPY]_TC` by steps of 0.1
```
```suggestion
* slow oscillations (1 oscillation per second or slower): this often occurs on large platforms and means that the attitude loop is too fast compared to the rate loop.
   - **Multicopter:** decrease `MC_[ROLL|PITCH]_P` by steps of 1.0
   - **Fixed-wing:** increase `FW_[R|P|Y]_TC` by steps of 0.1
```

> * fast oscillations (more than 1 oscillation per second): this is because the gain of the rate loop is too high -> decrease `MC_[ROLL|PITCH|YAW]RATE_K` by steps of 0.1

Reviewer: ```suggestion
* fast oscillations (more than 1 oscillation per second): this is because the gain of the rate loop is too high.
   - **Multicopter:** decrease `MC_[ROLL|PITCH|YAW]RATE_K` by steps of 1.0
   - **Fixed-wing:** decrease `FW_[RPY]R_P` by steps of 0.1
```

#### The auto-tuning sequence fails

If the drone was not moving enough during auto-tuning, the system identification algorithm might have issues to find the correct coefficients.
Increase the `MC/FW_AT_SYS_ID` by steps of 1 and trigger the auto-tune again.

#### The drone oscillates after auto-tuning

Due to effects not included in the mathematical model such as delays, saturation, slew-rate, airframe flexibility, the loop gain can be too high.

> To fix this, reduce the `MC_[ROLL|PITCH|YAW]RATE_K` parameters by steps of 0.1 until the oscillation stops.
> If a slow oscillation appears, reduce `FW_[RPY]R_P` by steps of 0.5.

Reviewer: ```suggestion
To fix this, follow the same steps described in [troubleshooting](#the-drone-oscillates-when-performing-the-testing maneuvers-prior-to-the-auto-tuning)
```

### FAQ

#### What frames types are supported?

> Autotuning is enabled for multicopter, fixed wing, and VTOL vehicles.

Reviewer: ```suggestion
Autotuning is enabled for multicopter, fixed wing, and hybrid VTOL fixed wing vehicles.
```

Author: Is this how we're referring to VTOLs now days? I think it is good to be precise, given that VTOL is not. But FWIW we don't make the distinction elsewhere and VTOL == hybrid VTOL fixed wing vehicles.

While it is not yet enabled for other frame types, in theory it an be used with any frame that uses a rate controller.

#### Does autotuning work for all supported airframes?

The mathematical model used by autotuning to estimate the dynamics of the drone assumes this it is a linear system with no coupling between the axes (SISO), and with a limited complexity (2 poles and 2 zeros).
If the real drone is too far from those conditions, the model will not be able to represent the real dynamics of the drone.

In practise, autotuning generally works well for fixed wing and multicopter, provided the frame is not too flexible.

#### How long does autotuning take?

Tuning takes 5s-20s per axis (aborted if tuning could not be established in 20s) + 2s pause between each axis + 4s of testing if the new gains are applied in air.

A multicopter must tune all three axes, and by default does not test the new gains in-air.

> The range is therefore between 19s (`5 + 2 + 5 + 2 + 5`) and 64s (`20x3 + 2x2`).

Author: @bresch You calculated the time for three axis as 3x time for 1 axis, and 3x space between axis. My thinking was that the space would actually be just between axes, and not afterwards (so 2x axis)? Or am I wrong here?

Reviewer: There is a 2s delay between the last axis and the in-air test too, so it's really 3x time for 1 axis + 3x space between axis

Author: Yeah, but by default you don't run the in-air test on multicopter, and this case is "the default multicopter case". I will fix up the default fixed wing case though.

By default a fixed wing vehicle tunes all three axes and runs tests the new gains in-air.

> The range is therefore between 23s (19+4) and 68s (64+4).

Author: ```suggestion
The range is therefore between  25s (`5 + 2 + 5 + 2 + 5 + 2 + 4`) and 70s (`20x3 + 3x2 + 4`).
```
